### PR TITLE
fix(SummarizationHook): preserve original SystemMessage during summarization

### LINKED ISSUE #4048: [BUG] 从源码上看，SummarizationHook（消息总结压缩）似乎会把system message也一起总结了，这合理吗？
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Current Behavior

com.alibaba.cloud.ai.graph.agent.hook.summarization.SummarizationHook，这个消息压缩hook，在总结历史消息的时候，把原system message也一起总结了，然后用总结后的文本作为新的system message，这不是很合理

### Expected Behavior

鉴于system message比较重要，建议把原system message提取出来，将总结后的文本拼接在原system message之后形成新的system message

### Steps To Reproduce

_No response_

### Environment

```markdown
Spring AI Alibaba version(s):1.1.0.0-RC2
```

### Debug logs

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by wade-microblueworld**
### Root Cause

When compressing conversation history, `SummarizationHook` was including **all** messages before the cutoff in the summarization list, including any `SystemMessage` instances. The original system prompt was sent to the LLM for summarization, and its content was replaced with the summarized version, losing critical instruction content.

### Fix

Preserve all `SystemMessage` instances before the cutoff index and prepend them at the top of the new message list, before the summary message. This ensures the original system instructions remain intact.

Fixes #4048

### CODE REVIEW

**Comment by yuluo-yx** [comment]
可以加下对应的单测

**Comment by wade-microblueworld** [comment]
已补充对应单测 @yuluo-yx :

- `SummarizationHookSystemMessageTest`:验证摘要触发时原始 SystemMessage 被保留在消息列表顶部、不会被纳入摘要输入,以及 `keepFirstUserMessage` 开关和未达阈值时的行为
- `SerializationUtilsMessageDeepCopyTest`:验证 UserMessage/SystemMessage/AssistantMessage(含 toolCalls、metadata)及消息列表的深拷贝语义

单测均为纯单元测试(mock ChatModel),不依赖外部 API,可直接在 CI 运行。另外顺带修复了合并 main 后 `deepCopyMessage` 与新版 Spring AI API 不兼容的编译问题。麻烦有空再看下,谢谢!

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/hook/summarization/SummarizationHook.java [hunk 1]
```diff
@@ -131,10 +131,18 @@ public AgentCommand beforeModel(List<Message> previousMessages, RunnableConfig c
 			}
 		}
 
+		// Preserve original SystemMessage(s) — they must not be summarized (fixes #4048)
+		List<Message> preservedSystemMessages = new ArrayList<>();
+		for (int i = 0; i < cutoffIndex; i++) {
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/hook/summarization/SummarizationHook.java:136]: The range [0, cutoffIndex) is iterated twice: once to collect preservedSystemMessages (lines 136-140) and again to build toSummarize (lines 143-148). These could be merged into a single loop, classifying each message as preserved, summarized, or skipped.

File: spring-ai-alibaba-graph-core/src/main/java/com/alibaba/cloud/ai/graph/utils/SerializationUtils.java [hunk 2]
```diff
@@ -155,6 +157,9 @@ public static Object deepCopyValue(Object value) {
 
 		// For other complex objects, try using Jackson serialization
 		// If it fails, return the original object (shallow copy)
+		if (value instanceof Message) {
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-graph-core/src/main/java/com/alibaba/cloud/ai/graph/utils/SerializationUtils.java:160]: On current main, UserMessage and SystemMessage are already handled inline in deepCopyValue() (lines 107-114) before this `instanceof Message` check is reached. The deepCopyMessage() branches for UserMessage (line 182) and SystemMessage (line 185) are dead code. The PR appears authored against an older version. Consider rebasing to remove the redundancy, or moving the `instanceof Message` check above the individual type checks to consolidate all Message handling in one place.

File: spring-ai-alibaba-graph-core/src/main/java/com/alibaba/cloud/ai/graph/utils/SerializationUtils.java [hunk 3]
```diff
@@ -165,4 +170,43 @@ public static Object deepCopyValue(Object value) {
...
+	 * manually reconstruct a new instance with copied content.
+	 */
+	private static Object deepCopyMessage(Message message) {
+		if (message instanceof UserMessage userMessage) {
+			return userMessage.copy();
+		}
+		if (message instanceof SystemMessage systemMessage) {
+			return systemMessage.copy();
+		}
+		if (message instanceof AssistantMessage assistantMessage) {
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-graph-core/src/main/java/com/alibaba/cloud/ai/graph/utils/SerializationUtils.java:188]: AssistantMessage.getMedia() returns List<Media>. Media objects are copied by reference inside the new ArrayList. If Media is mutable, this is not a true deep copy. Verify Media immutability or deep-copy each element.

File: spring-ai-alibaba-agent-framework/src/test/java/com/alibaba/cloud/ai/graph/agent/hook/messages/SummarizationHookSystemMessageTest.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		return SummarizationHook.builder()
+			.model(this.chatModel)
+			.maxTokensBeforeSummary(1)
+			.messagesToKeep(4)
+			.keepFirstUserMessage(keepFirstUserMessage)
+			.build();
+	}
+
+	@Test
+	void systemMessagePreservedAtTopWhenSummarizationTriggered() {
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-agent-framework/src/test/java/com/alibaba/cloud/ai/graph/agent/hook/messages/SummarizationHookSystemMessageTest.java:96]: Test class covers the happy path well but lacks edge-case tests: (1) no SystemMessage present — should not break summarization, (2) multiple SystemMessages — all should be preserved, (3) SystemMessage in recent messages (after cutoffIndex) — should pass through unchanged, (4) SystemMessage as the only message before cutoff — toSummarize becomes empty, createSummary returns 'No previous conversation.'

**Review by oss-sentinel-ai — COMMENTED** [review]
### Summary
Fixes a real bug: `SummarizationHook` now preserves SystemMessage(s) at the top of the conversation instead of feeding them into summarization. The core fix is correct and well-scoped — ordering preserved, no loss/duplication, edge cases (no system message, multiple system messages, below-threshold) handled safely; token-trigger logic unchanged. Tests cover the main preservation scenarios. Approve-worthy with minor comments.

### Findings
- **[Warning]** `SerializationUtils.java:160` — against current main the `UserMessage`/`SystemMessage` branches added in `deepCopyMessage()` are dead code (both types are already handled earlier in `deepCopyValue()`); the branch looks authored against an older main — consider rebasing
- **[Info]** `SummarizationHook.java:136` — the range `[0, cutoffIndex)` is iterated twice; the two loops can be merged
- **[Info]** `SerializationUtils.java:188` — `AssistantMessage.getMedia()` elements are copied by reference; verify `Media` immutability for a true deep copy
- **[Info]** test gaps: no SystemMessage present / multiple SystemMessages / SystemMessage in recent range / SystemMessage as only pre-cutoff message

### Suggestions
1. Rebase onto current main and drop the redundant `deepCopyMessage()` branches (or consolidate all `Message` handling in one place).
2. Add a no-SystemMessage regression test to guard the summarization path.

---
*Automated review by github-manager-bot*
